### PR TITLE
[proposal] Call Tracing Feature

### PR DESCRIPTION
**Pull request by andreilee**
Signed-off-by: Andrei Lee <andreilee@fb.com>

## Summary

Proposal for call tracing feature in Magma. Details in markdown file.

## Test Plan

N/A

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: docs/readmes/proposals/p005_call_tracing.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+### Design
+
+#### Orchestrator - API
+
+The API supports starting the call trace, ending the tracing session, and
+downloading the resulting capture.
+
+The API will be as follows:
+```
+GET      /networks/{network_id}/call_tracing/{trace_id}
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md]: I prefer calling this as /trace/{trace_params} which can include trace type 
or /tracing/call. Otherwise, for each type of trace, then we will be polluting the global namespace

File: docs/readmes/proposals/p005_call_tracing.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+In the second phase, features will be provided for filtering call traces by
+subscriber, protocol, or interface.
+
+## Phase 1
+
+#### User Flow
+
+For call tracing with Magma, we will provide the following user flow:
+1. User accesses NMS
+2. User selects either the gateway or subscriber that they would like to do
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md:77]: Why gateway or subscriber ? When is gateway only chosen ? Is that for interface tracing ?

File: docs/readmes/proposals/p005_call_tracing.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+The first phase of implementation will focus on bringing an MVP. Call traces
+will be able to be initiated from the NMS, and downloads of captures viewable
+through Wireshark after capture will be available.
+
+**Phase 2 - Call Trace Filtering**
+
+(December - January)
+
+In the second phase, features will be provided for filtering call traces by
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md:68]: Why is this moved to second phase ? Is there additional complexity in filtering the call traces based on the parameters ?

File: docs/readmes/proposals/p005_call_tracing.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        type: boolean
+```
+
+#### Orchestrator - ctraced
+
+On the gateway, the ctraced service will manage call tracing for the network.
+This service will be responsible for initiating call tracing sessions on
+gateways, and servicing downloads of call trace captures.
+
+Call traces will be stored in postgresql, and periodically, ctraced should
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md:187]: Let's have a default max purge timeout and user configurable purge timeout.

File: docs/readmes/proposals/p005_call_tracing.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    type: object
+    description: Full state object of a call trace
+    properties:
+      call_trace_available:
+        type: boolean
+```
+
+#### Orchestrator - ctraced
+
+On the gateway, the ctraced service will manage call tracing for the network.
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md:183]: To some extent, if we make this service generic, we could potentially use it for port mirroring features on behalf of the user. Though that is not the goal here. This might be in case user wants to capture traffic on actual physical interfaces or if we support capturing traffic samples on the port.

File: docs/readmes/proposals/p005_call_tracing.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  bytes trace_content = 1; // Max size of 4MB
+}
+```
+
+To perform call tracing, ctraced will use `tshark`, a terminal-based wireshark
+tool to do packet capture on the gateway.
+
+Call tracing will be initiated from the orc8r, and the packet-capture will be
+transferred back to orc8r from AGW via gRPC. Currently there is a limit for
+gRPC message sizes to 4MB, which limits the size of packet capture. Given that
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md:259]: Nice to know. Yes. I think setting a max capture size of 1-2MB should mostly suffice.

File: docs/readmes/proposals/p005_call_tracing.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+## Goals
+
+- Allow network operators to trace the control plane messaging for a subscriber
+- Allow network operators to trace control plane messaging for a specific
+protocol or interface
+
+## Additional Criteria
+
+These additional criteria are ideal to meet but are not necessary:
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md]: It will be nice if in phase 1, we can atleast implement a super simple interface in NMS, with just basic options of start, stop and downloading the pcap. Going to swagger and doing this is additional complexity which might be nice to avoid.

File: docs/readmes/proposals/p005_call_tracing.md [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+the packet capture should filter out data plane traffic, this should be
+sufficient for short tracing sessions, for an MVP.
+
+#### NMS - Interface
+
+A new section will be added to the NMS to initiate call traces,
+see past call traces, and download them.
+
+---
+
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md:269]: If we use tshark, the implementation across FWA, CWF, FEG should be similar. How do we plan to deploy the tshark binary ? Do we want to install it directly or deploy it as a container here ?

File: docs/readmes/proposals/p005_call_tracing.md [hunk 9]
```diff
@@ -0,0 +1,283 @@
+---
+id: p005_call_tracing
+title: Call Tracing for Troubleshooting
+hide_title: true
+---
+
+# Call Tracing for Troubleshooting
```
Comment (karthiksubraveti) [inline docs/readmes/proposals/p005_call_tracing.md]: minor nit: I guess since this includes subscriber and protocol, we can call it just Tracing support. I thought call tracing was subscriber specific

**Review by karthiksubraveti — COMMENTED** [review]
Thanks @andreilee for writing up this proposal. Proposal looks good. Left a few comments here.

File: docs/readmes/proposals/p005_call_tracing.md (hunk 2 above)
Comment (andreilee) [reply docs/readmes/proposals/p005_call_tracing.md:77]: It just does it for a gateway, but choosing it for a subscriber would then figure out the gateway for you.

Comment (andreilee) [reply docs/readmes/proposals/p005_call_tracing.md:77]: This would just be for phase 1, afterwards we would want more options

File: docs/readmes/proposals/p005_call_tracing.md (hunk 3 above)
Comment (andreilee) [reply docs/readmes/proposals/p005_call_tracing.md:68]: Short answer, yes. I'll add additional details to the proposal

File: docs/readmes/proposals/p005_call_tracing.md (hunk 5 above)
Comment (andreilee) [reply docs/readmes/proposals/p005_call_tracing.md:183]: Should I leave this section as-is?

File: docs/readmes/proposals/p005_call_tracing.md (hunk 8 above)
Comment (andreilee) [reply docs/readmes/proposals/p005_call_tracing.md:269]: Installed during gateway provisioning step

File: docs/readmes/proposals/p005_call_tracing.md (hunk 9 above)
Comment (andreilee) [reply docs/readmes/proposals/p005_call_tracing.md]: Sure, I'm good with that!

File: docs/readmes/proposals/p005_call_tracing.md (hunk 5 above)
Comment (karthiksubraveti) [reply docs/readmes/proposals/p005_call_tracing.md:183]: Yes. I think that sounds fine.

**Review by karthiksubraveti — APPROVED** [review]
LGTM. Thanks for writing this up @andreilee

**Comment by jpad-freedomfi** [comment]
@andreilee This proposal looks awesome. Very excited about these feature. 

Can we add the ability for user to specify non-standard ports for tracing? We have cases where OCS vendor specifies e.g. 3869 as port for Gy instead of 3868. So just want to make sure we have ability to trace in that scenario.

Perhaps have default ports, with ability for user to override default for each protocol if needed.

Also, I don't see mention of Federation Gateway in the proposal, only AGW and orc8r. Am I right to assume Gy/Gx/S6a would be captures would be on FeG interfaces? 

Thanks -Joey

**Comment by jpad-freedomfi** [comment]
Also, I assume the "GT=2" here is a typo and you mean "GY=2", is that right? 
Screen Shot 2020-11-16 at 9 03 07 AM
